### ISSUE TITLE
systemd-networkd seems to not retry renewing an expiring DHCP lease

### ISSUE THREAD

**Issue by adamnovak** [systemd/systemd#32045]
### systemd version the issue has been seen with

245.4-4ubuntu3.23

### Used distribution

Ubuntu 20.04.6 LTS

### Linux kernel version used

5.15.0-1056-aws

### CPU architectures issue was seen on

x86_64

### Component

systemd-networkd

### Expected behaviour you didn't see

If systemd-networkd goes to renew the DHCP lease an hour before expiry, and it fails, it should keep trying to renew the lease or get a new lease at regular intervals. This should be announced in the syslog logs.

### Unexpected behaviour you saw

In my syslog logs on my AWS instance, I saw this:
```
Mar 29 19:33:03 ip-172-31-10-143 systemd-networkd[2515543]: eth0: Could not set DHCPv4 address: Connection timed out
Mar 29 19:43:59 ip-172-31-10-143 systemd-networkd[2515543]: eth0: Failed
```

Then, an hour later, the instance dropped off the network, and remained offline until a reboot hours later, even though Amazon's DHCP server was *probably* functional for a nontrivial window while the machine was offline.

I think that systemd tried and failed to renew the lease once before it expired, and then sat quiescently as the lease actually expired and the instance lost Internet connectivity, without doing anything to try and get an IP again. I did *not* see anything in the logs indicating that systemd-networkd was making any DHCP discovery requests (successful or otherwise) while the system was offline.

### Steps to reproduce the problem

1. Have a machine get a DHCP lease.
2. When systemd goes to renew it before its expiration, make sure that it cannot contact the DHCP server at that particular time, but bring the DHCP server back online before the lease actually expires.
3. Observe a single renewal failure in the syslog, followed by the machine losing its IP apparently forever.

**Comment by yuwata** [comment]
Sorry, v245 is too old here. Please try to reproduce the issue with a recent release or the current git HEAD. Then, please provide the debugging logs of networkd.

**Comment by adamnovak** [comment]
It looks like the people supporting this build in this environment would be Canonical, and them [only through this month](https://ubuntu.com/about/release-cycle).

@yuwata You can confirm that this is not expected behavior on current systemd versions, and one *should* expect it to retry renewing a lease, right?

**Comment by yuwata** [comment]
Again, here is the upstream issue tracker, and v245 is too old here.
Please
- contact to your downstream, i.e. Ubuntu issue tracker or support forum, or
- try to reproduce the issue with a recent release (v254 or v255) or the current git HEAD, and provide debugging logs.

Without that, I cannot say anything about the issue. Sorry.

**Comment by CRCinAU** [comment]
I just came across this when trying to figure out why my DHCPv6 and PDs are expiring....

I get the following:
```
systemd-networkd[3000]: eth1: DHCPv4 address <ipv4>/24, gateway <gw> acquired from <gw>
systemd-networkd[3000]: eth1: DHCPv6 address <ipv6>/128 (valid for 29min 59s, preferred for 29min 59s)
systemd-networkd[3000]: eth1: DHCP: received delegated prefix <prefix>::/56
```

After 30 minutes, IPv6 just seems to stop working - and I have no IPv6 gateways anymore.

```
# rpm -qa | grep systemd-networkd
systemd-networkd-255.8-1.fc40.x86_64
```

**Comment by kugel-** [comment]
I've had this on `systemd 255 (255.4-1ubuntu8.4)`.

Then I tried `sudo networkctl renew nic0` without success, followed by `sudo networkctl forcerenew nic0` without success. Only after `sudo networkctl reconfigure nic0` I got a new DHCP lease.

On my job, I must use DHCP, and failing to renew means I cannot work remotely.

The situation is now recovered. Still, is there any information I can provide?

**Comment by kugel-** [comment]
At least the relevant journal logs
```
Nov 04 14:51:19 tmartitz-pc systemd-networkd[43464]: nic0: Could not set route: Die Wartezeit für die Verbindung ist abgelaufen
Nov 04 14:51:22 tmartitz-pc systemd-networkd[43464]: nic0: Failed
Nov 05 08:14:34 tmartitz-pc systemd-networkd[43464]: nic0: Configuring with /etc/systemd/network/51-haus.network.
Nov 05 08:14:34 tmartitz-pc systemd-networkd[43464]: nic0: DHCP lease lost
Nov 05 08:14:34 tmartitz-pc systemd-networkd[43464]: nic0: DHCPv6 lease lost
Nov 05 08:14:34 tmartitz-pc systemd-networkd[43464]: nic0: DHCPv4 address 172.17.92.148/24, gateway 172.17.92.1 acquired from 172.16.0.233
```

**Comment by adamnovak** [comment]
@yuwata Should this be re-opened since this has now been reproduced on v255? Or should @kugel- open a new issue?

**Comment by wkornewald** [comment]
With systemd 256 the issue can reproduced, too. Initially, a DHCPv6 address is retrieved. Once the DHCPv6 lease expires the IPv6 address is lost and no attempt is made to retrieve a new one. I've tried both `networkctl renew` and `forcerenew`, but nothing happens.

Currently, the only workaround is to periodically call `networkctl reconfigure`, but that causes the host to be unreachable for a brief moment.

**Comment by glennpratt** [comment]
Seeing similar, though still on 249:

```
root@myhost-1:~# journalctl -u systemd-networkd.service --since "3 days ago" | grep eth0
Feb 08 08:30:16 myhost-1 systemd-networkd[685]: eth0: Could not set route: Connection timed out
Feb 08 08:30:17 myhost-1 systemd-networkd[685]: eth0: Failed
```

# Manual renew error
```
root@myhost-1:~# networkctl renew eth0
Failed to renew dynamic configuration of interface eth0: Invalid argument
```

#### Reload no change
```
root@myhost-1:~# systemctl reload systemd-networkd.service
```

#### Restart fixes
```
root@myhost-1:~# systemctl restart systemd-networkd.service
root@myhost-1:~# journalctl -u systemd-networkd.service --since "1 hour ago" | grep eth0
Feb 10 16:51:53 myhost-1 systemd-networkd[2065406]: eth0: Link UP
Feb 10 16:51:53 myhost-1 systemd-networkd[2065406]: eth0: Gained carrier
Feb 10 16:51:53 myhost-1 systemd-networkd[2065406]: eth0: Gained IPv6LL
Feb 10 16:51:53 myhost-1 systemd-networkd[2065406]: eth0: DHCPv4 address 172.234.19.68/24 via 172.234.19.1
```

**Comment by zenlord** [comment]
Reproduceable with v257.3 on Archlinux. Every day at 3AM my WiFi access points optimize their settings, which last night triggered my laptop wlan0 to lose its carrier, and it did not recover the DHCPv6 acquired ip address on its own. I restarted systemd-networkd.service and all is well. Logs: 

```
mrt 02 03:01:27 ltx124 iwd[11956]: event: roam-scan,
mrt 02 03:01:27 ltx124 iwd[11956]: event: roam-scan,
mrt 02 03:01:30 ltx124 iwd[11956]: event: roam-info, bss: b6:fb:e4:aa:bb:cc, signal: -51, load: 2/255
mrt 02 03:01:30 ltx124 iwd[11956]: event: state, old: connected, new: ft-roaming
mrt 02 03:01:30 ltx124 kernel: wlan0: disconnect from AP b4:fb:e4:xx:yy:zz for new assoc to b6:fb:e4:aa:bb:cc
mrt 02 03:01:30 ltx124 systemd-networkd[11944]: wlan0: Lost carrier
mrt 02 03:01:30 ltx124 systemd-networkd[11944]: wlan0: DHCP lease lost
mrt 02 03:01:30 ltx124 systemd-networkd[11944]: wlan0: DHCPv6 lease lost
mrt 02 03:01:30 ltx124 kernel: wlan0: associate with b6:fb:e4:aa:bb:cc (try 1/3)
mrt 02 03:01:31 ltx124 kernel: wlan0: RX ReassocResp from b6:fb:e4:aa:bb:cc (capab=0x1111 status=0 aid=2)
mrt 02 03:01:31 ltx124 systemd-networkd[11944]: wlan0: Connected WiFi access point: XXXXXX (b6:fb:e4:aa:bb:cc)
mrt 02 03:01:31 ltx124 kernel: wlan0: associated
mrt 02 03:01:31 ltx124 iwd[11956]: event: state, old: ft-roaming, new: connected
mrt 02 03:01:31 ltx124 systemd-networkd[11944]: wlan0: Gained carrier
```

**Comment by AlexIn-Tech** [comment]
> With systemd 256 the issue can reproduced, too. Initially, a DHCPv6 address is retrieved. Once the DHCPv6 lease expires the IPv6 address is lost and no attempt is made to retrieve a new one. I've tried both `networkctl renew` and `forcerenew`, but nothing happens.
> 
> Currently, the only workaround is to periodically call `networkctl reconfigure`, but that causes the host to be unreachable for a brief moment.

I noticed the same thing with DHCPv6, I have systemd 255. 

I did a `tcpdump` on my client (Ubuntu 24.04 server) and what I can see is that my client will send a `RENEW` packet, receive a `REPLY` packet from the DHCPv6 server but it will do nothing, it will not renew the lease. It will continues to send `RENEW` packet and receive `REPLY` packet from DHCPv6 Server until T2 expires. Then it starts sending multiple `REBIND` packet until, eventually, my client sends a DHCPv6 `RELEASE`. 

The packets are valid. But my client (Ubuntu 24.04 server with systemd 255) doesn't apply the lease from the `REPLY` . 

I also tried `networkctl renew <interface>` and `networkctl forcerenew <interface>`  and it does nothing...

**Comment by CRCinAU** [comment]
I'm surprised that this is still a problem after a year - given how critical DHCP is to network operations....

**Comment by caizixian** [comment]
I think we possibly encountered the same issue, albeit in a more weird setup. Sometimes there are runaway processes that allocate too much and lead to swapping and then OOM. networkd seems to either be severely throttled or killed before being able to renew a DHCP lease, so even after the rogue processes exit, the server is still without a network connection.

**Comment by adamnovak** [comment]
@CRCinAU I think the high load/swapping/OOM kill part is probably actually required to reproduce the issue, so it won't happen for most people, so it ends up being a niche problem. I didn't describe it in my report at the top, but I think the systems I'm seeing this issue with are also plausibly experiencing periods of high load/memory pressure that are consistent with that being a required ingredient to trigger the problem. What I wrote as "steps to reproduce" was a guess based on the logging I was looking at; I didn't actually build and test a setup to reliably trigger the issue on demand.

**Comment by nijave** [comment]
> [@CRCinAU](https://github.com/CRCinAU) I think the high load/swapping/OOM kill part is probably actually required to reproduce the issue, so it won't happen for most people, so it ends up being a niche problem. I didn't describe it in my report at the top, but I think the systems I'm seeing this issue with are also plausibly experiencing periods of high load/memory pressure that are consistent with that being a required ingredient to trigger the problem. What I wrote as "steps to reproduce" was a guess based on the logging I was looking at; I didn't actually build and test a setup to reliably trigger the issue on demand.

Anecdotally, I've seen similar behavior. System gets overloaded, loses lease. Load drops but system never regains lease and requires manual intervention.

In these situations, I don't see any indication systemd was impacted so maybe the high load led to some sort of timeout being hit.

**Comment by CRCinAU** [comment]
Back when I commented here, mine was mostly under idle scenarios - so I doubt its required to have high load to suffer from it.

**Comment by kugel-** [comment]
On my system it usually happens after overnight suspend-to-ram, but sometimes also intermittently during normal usage. No high load involved.

**Comment by alexcorriedt** [comment]
We're also observing this - usually over periods of unusually high load that sync up with a dhcp renewal. `systemd-networkd` has one attempt, if that fails there are no retries or other attempts until the VM drops off networking and has to be restarted to bring it back up. 

sar load logging
```
20:35:01      runq-sz  plist-sz   ldavg-1   ldavg-5  ldavg-15   blocked
22:35:01           41      2827     11.52     13.19     17.25         0
23:04:16            5      3066    673.75    625.89    443.64       195
23:05:01           21      2778    362.71    550.56    426.79         3
23:15:01           16      2934      3.18     78.92    227.44         1
23:25:01           14      2896      1.19     11.97    120.03         2
23:35:01           16      2905      1.85      3.23     63.77         1
```

systemd-networkd
```
Jun 18 23:04:15 vm systemd-networkd[962]: ens64: Could not set DHCPv4 address: Connection timed out
```

The load is clearly not ideal, but it is recoverable as long as networking doesn't go down entirely because of it. We observe no retries, and I don't have a good way to make it retry short of hitting "reboot" on the cloud provider. 

Appreciate any suggestions from anyone to help ameliorate this and force a retry either through systemd config or other means.

**Comment by CRCinAU** [comment]
> Appreciate any suggestions from anyone to help ameliorate this and force a retry either through systemd config or other means.

Only solution I've found over the last year+ is to uninstall systemd-networkd and use Network Manager.

**Comment by kayrus** [comment]
@CRCinAU this is not an option in flatcar

**Comment by CRCinAU** [comment]
> @CRCinAU this is not an option in flatcar

This is the problem when building stuff on top of an unstable house of cards. Given this is at least a year old and not fixed yet, I'm not really convinced that the systemd folk have any interest in having a stable networking component.

Try using something proven like maybe including and launching dhclient in your projects - as you probably shouldn't hold your breath for this to be fixed - and its a major, disruptive, and basic functionality breaking issue that has gone unfixed for a long time.

**Comment by adamnovak** [comment]
This being a major disruptive issue for a lot of people also might mean fixing it is our job.

I unfortunately couldn't find any developer-facing documentation for systemd-networkd that explains how it is designed or what parts are supposed to be doing what at any given time. If some exists, someone should reference it here.

The DCHP client tests appear to live here: https://github.com/systemd/systemd/blob/d14cfb6e82da4454c14a2f009ca68ef76ae04417/test/test-network/systemd-networkd-tests.py#L7329

There's a couple points at which successful renewal is tested, but no point at which a renewal *failure* is tested, and all the expiration/renewal tests appear to be using real time waits instead of somehow mocking the clock to force expiration. So it's plausible that the real root cause of the problem is insufficient testing in stressful/abnormal conditions, and if someone wrote a renewal failure test this would just reproduce every time.

Some relevant-looking code is here:
https://github.com/systemd/systemd/blob/d14cfb6e82da4454c14a2f009ca68ef76ae04417/src/network/networkd-dhcp4.c#L1789-L1806
(Note how if told to `dhcp4_renew()` in any state other than `DHCP_STATE_BOUND`, we `sd_dhcp_client_interrupt_ipv6_only_mode()` instead of `sd_dhcp_client_send_renew()`ing. This in turn reinitializes and restarts the client, which *probably* won't get it stuck because the first thing it is likely to try and do is get a lease, but it does look like trying to renew when already in `DHCP_STATE_RENEWING` will produce this somewhat astonishing result.)

And some more relevant-looking code, which looks like the one true timeout handler for the DHCP client, is here:
https://github.com/systemd/systemd/blob/d14cfb6e82da4454c14a2f009ca68ef76ae04417/src/libsystemd-network/sd-dhcp-client.c#L1251-L1388

When there's a timeout event, it checks to see what it was meant to be doing based on the client state, figures out what the next timeout should be, schedules a timeout event for that time in the future, and then again based on the state it decides what sort of packets to resend or other action to take to deal with there having been a timeout.

I see a couple potential problems:

1. Any sort of internal error is responded to with a `goto error`, and the error handling logic, in the case where it hasn't exhausted its retries, just stops the client. So if we're in `DHCP_STATE_REBOOTING` when a timeout occurs, and `client_start()` has some kind of transient error, we end up in `DHCP_STATE_STOPPED`. I don't think any other system component is responsible for kicking the DHCP client on an interface out of `DHCP_STATE_STOPPED` when it oughtn't be there. As a user of the DHCP client, I would expect it not to ever reach the stopped state, after I've requested to start it, until I requested to stop it. So perhaps the error handling logic should be changed to restart the client instead.
2. Future retries rely entirely on the reliability of the event/timer system. Are we sure that `event_reset_time_relative()` events can never be lost under any circumstances? It looks like it is implemented by getting the current time, computing the target future time, and then setting the event to fire at that target time; if the process sleeps long enough between checking the time and scheduling the event that the target time is in the past by the time the event is actually registered, are we sure the event will actually fire? If any of the codepaths end up trying to cancel the timeout event (especially asynchronously), are we sure they will always manage to do so before the event can be delivered?

**Comment by adamnovak** [comment]
It looks like https://github.com/systemd/systemd/issues/33278 might also be a duplicate of this, though on a too-old systemd version. @yuwata suggested that the underlying cause might be that, under high load, "networkd could not receive response from the kernel within a certain time span", but didn't specify what kind of response that might be. But if we could find out what message/response is suspected, we could try and add some kind of handling for a missed deadline.

I'm not able to find in the code the source of the "Could not set DHCPv4 address" message; it's possible the string has been changed in the current systemd version?

**Comment by alexcorriedt** [comment]
> It looks like [#33278](https://github.com/systemd/systemd/issues/33278) might also be a duplicate of this, though on a too-old systemd version. [@yuwata](https://github.com/yuwata) suggested that the underlying cause might be that, under high load, "networkd could not receive response from the kernel within a certain time span", but didn't specify what kind of response that might be. But if we could find out what message/response is suspected, we could try and add some kind of handling for a missed deadline.
> 
> I'm not able to find in the code the source of the "Could not set DHCPv4 address" message; it's possible the string has been changed in the current systemd version?

With regards to the error logging all I can find that is a *maybe* is https://github.com/systemd/systemd/blob/7478fbb0af13dfd787b01cf1c9488bfcae07d705/src/hostname/hostnamectl.c#L585 which looks to be the correct format to match `eth0: Could not set DHCPv4 address: Connection timed out`

However it's not where I'd have guessed the error handling for setting the dhcp address would be so I could be well out.

**Comment by rajibade** [comment]
@adamnovak @yuwata Any chance there just isn't any type of proper retry mechanism here. From the other comments and issues, seems it could be a plethora of different missed deadlines and messages, either due to a kernel in high load situations or the cases of idle timeouts. Seems the renew checks could do some form of retry on fails.

**Comment by adamnovak** [comment]
@rajibade I think that's generally true? It's all based around a few timers, and if somehow whatever was supposed to happen at one of the timers doesn't happen, in a way that doesn't set the timer again, there's no second level of supervision that ever detects the bad state and tries to drive the system towards the good state.

But I'm not really sure what a "proper" retry mechanism would look like; there would need to be... a timer, to decide when to check if things need retrying. And then we're back where we started, with it being possible for the timers to be in a bad state.

**Comment by nijave** [comment]
> [@rajibade](https://github.com/rajibade) I think that's generally true? It's all based around a few timers, and if somehow whatever was supposed to happen at one of the timers doesn't happen, in a way that doesn't set the timer again, there's no second level of supervision that ever detects the bad state and tries to drive the system towards the good state.
> 
> But I'm not really sure what a "proper" retry mechanism would look like; there would need to be... a timer, to decide when to check if things need retrying. And then we're back where we started, with it being possible for the timers to be in a bad state.

A global deadline/timeout to obtain a lease that causes it to crash so supervision can be delegated elsewhere (systemd)?

**Comment by CRCinAU** [comment]
> But I'm not really sure what a "proper" retry mechanism would look like; there would need to be... a timer, to decide when to check if things need retrying. And then we're back where we started, with it being possible for the timers to be in a bad state.

Thankfully, DHCP is defined in an RFC.

https://www.rfc-editor.org/rfc/rfc2131#section-4.4.5

In brief:

   In both RENEWING and REBINDING states, if the client receives no
   response to its DHCPREQUEST message, the client SHOULD wait one-half
   of the remaining time until T2 (in RENEWING state) and one-half of
   the remaining lease time (in REBINDING state), down to a minimum of
   60 seconds, before retransmitting the DHCPREQUEST message.

**Comment by rajibade** [comment]
> > But I'm not really sure what a "proper" retry mechanism would look like; there would need to be... a timer, to decide when to check if things need retrying. And then we're back where we started, with it being possible for the timers to be in a bad state.
> 
> Thankfully, DHCP is defined in an RFC.
> 
> https://www.rfc-editor.org/rfc/rfc2131#section-4.4.5
> 
> In brief:
> 
> In both RENEWING and REBINDING states, if the client receives no response to its DHCPREQUEST message, the client SHOULD wait one-half of the remaining time until T2 (in RENEWING state) and one-half of the remaining lease time (in REBINDING state), down to a minimum of 60 seconds, before retransmitting the DHCPREQUEST message.

Which thankfully is what is observably being done in the normal operation. I suppose "retry" isn't the right word here.

**Comment by kayrus** [comment]
Just checking in, any updates on this issue? Is there a way to prioritize or highlight it for the maintainers?"

**Comment by rajibade** [comment]
Can't remember if this one deals with the timeout but I have a PR here #38429. Hoping to get some feedback soon

**Comment by adamnovak** [comment]
After reading the discussion at https://github.com/systemd/systemd/pull/38429, I now think the offending timeouts might not be in the DHCP client at all. The problem might be the timeout [controlled by the `SYSTEMD_NETLINK_DEFAULT_TIMEOUT` environment variable](https://systemd.io/ENVIRONMENT/), which defaults to 25 seconds.

It seems like the workaround would be to `systemctl edit systemd-networkd.service` to create an override file for the systemd unit, and put in it:

```
[Service]
Environment="SYSTEMD_NETLINK_DEFAULT_TIMEOUT=infinity"
```

See also:
* https://github.com/systemd/systemd/issues/25441
* https://gist.github.com/raggi/1f8d0b9f45c5b62e7131b03e6e2ffe68

**Comment by adamnovak** [comment]
I've updated my original offending system to Ubuntu 24.04, which ships systemd 255, which is the oldest release that actually has that environment variable for controlling the timeout. So I'm trying out that possible workaround, but I'm not sure I'll be able to tell if it works, because I also have a cron job set up to kick the system back onto the network when it drops off.

**Comment by kugel-** [comment]
@adamnovak does not help on my system (which is Ubuntu 24.04). I'm still experiencing no-renew. It's a nightmare, it often means I cannot connect to my workstation from home office.

**Comment by adamnovak** [comment]
@yuwata I think now we have at least one report form @kugel- that the workaround you wanted tried in https://github.com/systemd/systemd/pull/38429#issuecomment-3224662583 with `SYSTEMD_NETLINK_DEFAULT_TIMEOUT` doesn't actually solve the problem in all cases.

(I haven't myself noticed the problem recurring with the workaround and systemd 255, in about a month of syslog logs or AWS status checks, which suggests that it might be working for me or that I might not have high load lately.)

**Comment by kayrus** [comment]
Hi. We still experience this issue. Do you have updates?

**Comment by raggi** [comment]
Dup of https://github.com/systemd/systemd/issues/25441

**Comment by benjimin** [comment]
I recently experienced this scenario with an AWS EC2 VM:

(DHCP lease has lifetime 1h, T1 30min, T2 52min:30. Single network interface for external connectivity including to VM hypervisor IMDS virtual endpoint. Application is for batch data transfer and processing.)

- ~1300-1355ish: thrashing shown by CPU utilisation metrics (5min granularity), presumably due to python app memory leak.
- 1356:26-:29: networkd failed to set DHCPv4 address (connection timed out). I'm guessing the thrashing stalled networkd while letting the network buffer overflow (dropping packets containing the lease renewal acceptance).
- 1358:45 kernel OOM killed python (presumably restoring system load to normal)
- 1358:48-59:18 timedated run (by snapd via dbus), which I'm guessing may use NTP to correct for any clock drift resulting from clock ticks missed during thrashing.
- 1428: logs show processes such as ssm-agent can still retrieve short-term credentials from IMDS
- ~1440: network traffic ceases according to metrics (5min granularity)
- 1445: EC2 "instance" (VM network connectivity) status checks start failing (1min granularity), but other EC2 status checks (host system hardware status and host's connection to EBS network storage for VM filesystem) remain ok.
- 1458:47: logs show processes such as ssm-agent failing to even access IMDS (network is unreachable)
- For several days until manual restart, status never restored. Cron jobs continue running on VM, but write logs indicating no network connectivity.

It might be worth making a unit test that tries to replicate this scenario, by having the mock DHCP server initially fail to respond to a renewal, and then somehow accelerating the clock to check that networkd performs a retry later?
